File: services/rust/salvo/src/routes/middleware/log.rs
use salvo::{Depot, FlowCtrl, Request, Response};
use tracing::debug;

#[salvo::handler]
pub async fn log_middleware(
    req: &mut Request,
    depot: &mut Depot,
    res: &mut Response,
    ctrl: &mut FlowCtrl,
) {
    debug!(method = %req.method(), path = %req.uri(), "Incoming request");

    ctrl.call_next(req, depot, res).await;
}

File: services/python/fastapi/src/routes/middleware/log.py
import structlog
from fastapi import Request, Response
from collections.abc import Awaitable, Callable


async def log_middleware(
    request: Request,
    call_next: Callable[[Request], Awaitable[Response]],
) -> Response:
    structlog.get_logger().debug(
        "Incoming request", method=request.method, path=request.url.path
    )

    return await call_next(request)
